fhirengine/src/main/java/com/google/fhirengine/db/impl/dao/LocalChangeUtils.kt

```diff
@@ -0,0 +1,122 @@
+        val type: LocalChange.Type
+        val diff: String
+        when (second.type) {
+            LocalChange.Type.UPDATE -> when {
+                first.type.equals(LocalChange.Type.UPDATE) -> {
+                    type = LocalChange.Type.UPDATE
+                    diff = mergePatches(first.diff, second.diff)
+                }
+                first.type.equals(LocalChange.Type.INSERT) -> {
+                    type = LocalChange.Type.INSERT
+                    diff = applyPatch(first.diff, second.diff)
+                }
+                else -> {
+                    throw IllegalArgumentException(
+                            "Cannot merge local changes with type ${first.type} and ${second.type}."
+                    )
+                }
+            }
+            LocalChange.Type.DELETE -> {
+                type = LocalChange.Type.DELETE
+                diff = ""
+            }
+            else ->
+                throw IllegalArgumentException(
+                    "Cannot merge local changes with type ${first.type} and ${second.type}."
+                )
+        }
+        return LocalChange(
+            id = 0,
+            resourceId = second.resourceId,
+            resourceType = second.resourceType,
+            type = type,
+            diff = diff
+        )
+    }
+
+    /**
+     * Update a JSON object with a JSON patch (RFC 6902).
+     */
+    @JvmStatic
+    private fun applyPatch(resourceString: String, patchString: String): String {
+        val objectMapper = ObjectMapper()
+        val resourceJson = objectMapper.readValue(resourceString, JsonNode::class.java)
+        val patchJson = objectMapper.readValue(patchString, JsonPatch::class.java)
+        return patchJson.apply(resourceJson).toString()
+    }
+
+    /**
+     * Merge two JSON patch strings by concatenating their elements into a new JSON array.
+     */
+    @JvmStatic
+    private fun mergePatches(firstPatch: String, secondPatch: String): String {
+        val objectMapper = ObjectMapper()
+        // TODO: validate patches are RFC 6902 compliant JSON patches
+        val first = objectMapper.readValue(firstPatch, JsonPatch::class.java)
+        val updater = objectMapper.readerForUpdating(first)
+        val merged = updater.readValue(secondPatch, JsonPatch::class.java)
+        return merged.toString()
+    }
+
+    /**
+     * Calculates the JSON patch between two [Resource]s.
+     */
+    @JvmStatic
+    internal fun diff(parser: IParser, source: Resource, target: Resource): String {
+        val objectMapper = ObjectMapper()
+        val sourceJson = objectMapper.readValue(
+            parser.encodeResourceToString(source),
+            JsonNode::class.java
+        )
+        val targetJson = objectMapper.readValue(
+            parser.encodeResourceToString(target),
+            JsonNode::class.java
+        )
+        val jsonDiff = JsonDiff.asJson(sourceJson, targetJson)
```

Comment:
I actually think this might look better and more readable if you inline source and target jsons?

```suggestion
        val jsonDiff = JsonDiff.asJson(
            objectMapper.readValue(
                parser.encodeResourceToString(source),
                JsonNode::class.java
            ),
            objectMapper.readValue(
                parser.encodeResourceToString(target),
                JsonNode::class.java
            )
        )
```